### PR TITLE
implement udev "eternal" tags idea

### PR DESCRIPTION
**Pull request by poettering**
My proposal to address the kernel breakage that causes #7587.

### CODE REVIEW

**Comment by poettering** [comment]
I only tested this lightly so far, would be thankful for testing on affected systems.

**Comment by dtor** [comment]
Just for the record, I object to all of this and the way you propose to handle this. If you do not want bind/unbind, filter them out, do not pass it on to rules, and you will be in exactly the same position as you were before 4.12 kernel. Do not require consumers to change the rukes to "ACTION!="remove". This is a recommendation that will hurt in the long run.

That is my recommendation for now. Whether you decide to implement it or not is up to you obviously.

File: NEWS [hunk 1]
```diff
@@ -2,6 +2,79 @@ systemd System and Service Manager
...
+        * KERNEL API BREAKAGE I: Please note that Linux 4.12 introduced two new
+          uevents "bind" and "unbind", which break all userspace that is not
+          prepared to handle them (which is actually *most* of the packages
+          that process or monitor kernel devices, in all current
+          versions). It's highly recommended to update to systemd/udev 240 when
+          running kernels 4.12 or newer. Moreover, this change in behaviour
+          requires additional changes in almost all other software installing
+          udev rules files, monitoring or enumerating devices with libudev or
+          otherwise processing uevents. Note that this breakage is not fault of
+          systemd or udev, but exclusively caused by an incompatible kernel
```
Comment (dtor) [inline NEWS:14]: Yeah, it is though, as you did merge my patch enabling handling these 2 new actions in udev.

File: NEWS [hunk 2]
```diff
@@ -2,6 +2,79 @@ systemd System and Service Manager
...
+            udev_device_has_tag() API which checks if the tag ever existed on
+            the device, following the API concept redefinition explained
+            above).
+
+          We are very sorry for this breakage and the requirement to update
+          packages using these interfaces. We'd again like to underline that
+          this is not caused by systemd/udev changes, but result of a kernel
+          behaviour change that was not carefully enough discussed and
+          tested. Also note that more recent post-4.12 kernels generate the
+          "bind" and "unbind" in more subsystems than the earlier post-4.12
```
Comment (dtor) [inline NEWS]: No, that is not true either. The bind/and unbind form the beginning was generated for all drivers.

Comment (poettering) [reply NEWS]: hmm, i assumed i2c and stuff didn't generate this but do now. but if i was mistaken will remove this.

File: NEWS (hunk 1 above)
Comment (poettering) [reply NEWS:14]: Well, my assumption was that you tested this. udev is just the userspace wrapper around kernel behaviour, and was less than well maintained in the recent past. I am not sure I am willing to take the blame for this breakage, in particular as your patch simply changed udev to handle these two events like they were processed in older udev versions that didn't validate kernel action types. So, dunno. You broke udev, and your patch didn't really make things better but not much worse either...

File: NEWS (hunk 2 above)
Comment (poettering) [reply NEWS]: force pushed a new version that drops this line.

**Comment by dtor** [comment]
> i am sorry but your proposal can't work. If we filter out "bind" entirely, then all programs waiting for devices to become usable will never notice when they do, as they will only get the "add" at which they are useless still.

I believe it would work as the world was chugging along with kernels before 4.12 and even after 4.12 but with older versions of udev/systemd. We'd still have the same issues that we had before with some device attributes being not ready at the "add" time because they are provided by a driver, and not subsystem, but these issues existed before and are quite minor. Resurrecting them seems much lesser pain than having tags disappeared.

> Anyway, make a better recommendation. Simply ignoring bind/unbind entirely in userspace doesn't really work. Also, I'd much prefer a proper solution that works with released kernels since 4.12, than a temporary stop-gap, that doesn't work really at all...

I think stop-gap returning us to pre-4.12 days is actually quite sensible as I explained above. Longer term I expressed my idea of having better tag management for different actions and avoiding resetting the state unless action explicitly calls for it. I will try looking into it but I won't commit to a particular timeline unfortunately ;(

File: NEWS (hunk 1 above)
Comment (dtor) [reply NEWS:14]: I tested it on my set up and my use case. I'll take the blame on udev side if you wish ;)

**Comment by sourcejedi** [comment]
> We'd still have the same issues that we had before with some device attributes being not ready at the "add" time because they are provided by a driver, and not subsystem, but these issues existed before and are quite minor.

@dtor assuming we are having significant issues in comparison.  I haven't double-checked, but I'm guessing "bind" breaks enough things that it's worth thinking about a revert as you suggest.

I found this fairly attractive.  Maybe we break someone's custom rule that they needed to make their input device work reasonably (sensitivity, whatever).  But if *you* haven't heard about people deploying that in large installations, maybe we get away with it.

However, my reservation about a revert was the *firmware loading* stuff.  It sounded quite important, and it was important enough for you to bring it up.  For this case, I would want to have a bit more solid confirmation that no-one has deployed such a solution that depends on "bind" events in udev.  In practice I don't really know how we would make that evaluation, how we could be confident, if we think there's a risk some touchpads will stop operating?

File: NEWS [hunk 3]
```diff
@@ -2,6 +2,76 @@ systemd System and Service Manager
...
+          that the "tags" concept has been redefined this way, frequently it's
+          important to determine which tags are the ones applied by the most
+          recent uevent/database update, in order to discern them from those
+          originating from earlier uevents/database updates. To accommodate for
+          this a new automatic property CURRENT_TAGS has been added that works
+          similar to the existing TAGS property but only lists tags in effect
+          from the most recent uevent/database update. Similar, the
+          libudev/sd-device API has been updated with new functions to
+          enumerate these 'current' tags, in addition to the existing APIs to
+          enumerate the 'eternal' ones.
```
Comment (keszybz) [inline NEWS:36]: Looking at this again after some time, I think the naming of "current" and "eternal" is likely to be misunderstood. Here, "current" is not current, it's just the latest batch. I think we need to come up with some different naming.

**Comment by poettering** [comment]
So I rebased this now, and posted a new version as #16853. It's so heavily updated, that I made it a new PR. Closing this one hence.

**Pull request closed without merging**